### PR TITLE
Feature/page orientation

### LINKED ISSUE #173: import segmentation respecting `@orientation`
When I open existing PAGE-XML files with stand-alone LAREX, it does not deskew according to the `@orientation` angle annotated.

It's probably just a little more work manually (where you always have to draw polygons when you could live with rectangles), but for the automatic parts I guess these pages won't work well, will they?

I remember dimly that preprocessing is _prior_ to GT annotation work in the OCR4all process model. So will rotation (and internal coordinate transformation) ever be supported here in LAREX?

### PR DESCRIPTION
**Pull request by chaddy314**
Larex now deskews images and regions according to the @orientation angle annotated.
/PcGts/Page/@orientation is read from pageXML and is added to the PageAnnotations send to the client.
The rotation of each region will be executed during controller.loadAnnotations() and rotated back on export.
closes #173

### CODE REVIEW

File: src/main/java/de/uniwue/web/io/PageXMLReader.java [hunk 1]
```diff
@@ -217,4 +220,26 @@ public static PageAnnotations loadPageAnnotationsFromDisc(File pageXMLInputPath)
...
 		return segResult;
 	}
+
+	/**
+	 * Read the Orientation of a PageXML Document
+	 *
+	 * @param pagexml Path to the PageXML file
+	 * @return orientation skew angle of page element
+	 */
+	public static double getPageOrientation(Document pagexml) {
```
Comment (maxnth) [inline src/main/java/de/uniwue/web/io/PageXMLReader.java:230]: We should use prima-core-libs whenever possible so I propose replacing this function with
```Java
public static double getOrientation(Page page){
	if(page.getAttributes().get("orientation") != null && page.getAttributes().get("orientation").getValue() != null){
		return Double.parseDouble(page.getAttributes().get("orientation").getValue().toString());
	}
	return 0.0;
}
```

File: src/main/java/de/uniwue/web/io/PageXMLWriter.java [hunk 2]
```diff
@@ -163,13 +164,36 @@ public static Document getPageXML(PageAnnotations result, String pageXMLVersion)
...
 
+	/**
+	 * Writes the Orientation to a PageXML Document
+	 *
+	 * @param doc PageXML Document
+	 * @param orientation orientation angle in degree
+	 * @return modified Document
+	 */
+	private static Document setPageOrientation(Document doc, double orientation) {
+		Element rootElement = doc.getDocumentElement();
```
Comment (maxnth) [inline src/main/java/de/uniwue/web/io/PageXMLWriter.java:188]: See above:
Content of the function (which should digest `Page page, double orientation`) should be replaced with something like this:
```Java
page.getAttributes().get("orientation").setValue(new DoubleValue(orientation));
```

File: src/main/webapp/resources/js/viewer/controller.js [hunk 3]
```diff
@@ -688,6 +719,15 @@ function Controller(bookID, accessible_modes, canvasID, regionColors, colors, gl
 	}
 
 	this.exportPageXML = function (_page = _currentPage) {
+		if(_segmentation[_currentPage] != null) {
```
Comment (maxnth) [inline src/main/webapp/resources/js/viewer/controller.js:722]: Looks to be the same code as in lines 493-502. Should be encapsulated inside function.

**Review by maxnth — COMMENTED** [review]
Added some smaller comments.
Especially the non-"prima-core-libs" reading and writing should be avoided.

File: src/main/webapp/resources/js/viewer/controller.js [hunk 4]
```diff
@@ -204,15 +205,18 @@ function Controller(bookID, accessible_modes, canvasID, regionColors, colors, gl
 				} else {
 					this._requestEmptySegmentation();
 				}
-				pages.forEach(page => { _gui.addPageStatus(page, PageStatus.SERVERSAVED)});
+
```
Comment (maxnth) [inline src/main/webapp/resources/js/viewer/controller.js:207]: Why was this removed? After removing the icon for pages with a saved annotation isn't shown when the page hasn't been selected yet. Am I missing something?

File: src/main/webapp/resources/js/viewer/viewer.js [hunk 5]
```diff
@@ -474,6 +474,41 @@ class Viewer {
...
 
+	rotateImage(orientation, point) {
+		if (point == null) {
+			point = this._image.bounds.center;
+		}
+		this._image.rotate(orientation, point);
+		/*		let newBounds = this._image.bounds.clone();
+
+                let trueCenter = new paper.Point(newBounds.topLeft.x + (newBounds.topRight.x - newBounds.topLeft.x)/2, newBounds.topLeft.y + (newBounds.bottomLeft.y - newBounds.topLeft.y)/2);
+                return trueCenter;*/
```
Comment (maxnth) [inline src/main/webapp/resources/js/viewer/viewer.js:485]: Can the commented out code be deleted or is it useful for debugging or sth else?

**Comment by maxnth** [comment]
Running `Segment` on a page currently resets any orientation read from the file. As LAREX segmentation should only affect the Layout this probably isn't the expected behavior for most users (?)

https://user-images.githubusercontent.com/33344081/122058720-316af080-cdec-11eb-8b68-2e043b51ffdf.mp4

File: src/main/webapp/resources/js/viewer/controller.js (hunk 4 above)
Comment (chaddy314) [reply src/main/webapp/resources/js/viewer/controller.js:207]: No, you're absolutely right. Added some functions to the loop which I later removed again and accidentally removed outer loop with it.

File: src/main/webapp/resources/js/viewer/viewer.js (hunk 5 above)
Comment (chaddy314) [reply src/main/webapp/resources/js/viewer/viewer.js:485]: Can be deleted.

**Comment by maxnth** [comment]
Saving (via `Ctrl+S` or the `Save Result` button) throws the following error:
```javascript
controller.js:2168 Uncaught ReferenceError: i is not defined
    at controller.js:2168
    at Array.forEach (<anonymous>)
    at Controller.unrotateSegments (controller.js:2167)
    at Controller.exportPageXML (controller.js:705)
    at HTMLDocument.KeyInput.document.onkeydown (keyInput.js:279)
```

**Comment by maxnth** [comment]
Saving should now work.

**Comment by chaddy314** [comment]
Rebased to current dev version(57dd0d504413f3aa5193540ee586854f4b808e4b).
Saving should now work again, as well as proper derotation of segments.

**Comment by maxnth** [comment]
Orientation doesn't reset on page change.

With this project structure:
image
The second page is rotate 90 degrees even though no XML exists:
image

**Comment by maxnth** [comment]
This branch produces worse segmentation results for pages which were rotated.

The following segmentatione results were produces by LAREX segmentation with the same settings on the same pages:
a) Rotated
image
b) Not rotated
image

I guess we should also rotate the images in the backend before running the segmentation algorithm (or at least make it user configurable)

**Comment by maxnth** [comment]
I guess rotating directly when setting image 
https://github.com/OCR4all/LAREX/blob/005aaea52616ec869b776895b5a28d9d11c37e29/src/main/webapp/resources/js/viewer/controller.js#L209
instead of afterwards https://github.com/OCR4all/LAREX/blob/005aaea52616ec869b776895b5a28d9d11c37e29/src/main/webapp/resources/js/viewer/controller.js#L234
would be an improved as the non-rotated image wouldn't appear before rapidly being rotated.

**Comment by chaddy314** [comment]
> Orientation doesn't reset on page change.
> 
> With this project structure:
> image
> The second page is rotate 90 degrees even though no XML exists:
> image

fixed in latest commit

**Comment by maxnth** [comment]
All problems / bugs but the one regarding segmentation results seem to be resolved and ready for a merge into `dev`. 
While the open problem is IMO a very important one and one which has to be absolutely addressed before pushing the `dev` branch to `staging` I'll merge this PR already to ease testing this in combination with the other new features.
